src/WCSimIBDGen.cc

```diff
@@ -0,0 +1,258 @@
+#include "WCSimIBDGen.hh"
+#include "WCSimDetectorConstruction.hh"
+#include "WCSimPrimaryGeneratorAction.hh"
+#include "WCSimPrimaryGeneratorMessenger.hh"
+
+#include "json.hpp"
+
+#include <CLHEP/Units/PhysicalConstants.h>
+#include <CLHEP/Units/SystemOfUnits.h>
+#include <G4ThreeVector.hh>
+#include <G4Types.hh>
+#include <G4ios.hh>
+#include <fstream>
+
+// Set json to nlohmann::json
+using json = nlohmann::json;
+
+WCSimIBDGen::WCSimIBDGen(WCSimDetectorConstruction *myDC) : myDetector(myDC) {}
+
+WCSimIBDGen::~WCSimIBDGen() {
+    // Delete things here
+}
+
+void WCSimIBDGen::ReadSpectrumFromDB(G4String spectrum_name, std::string model_name) {
+    // Open the database
+    std::ifstream spectrum_json(spectrum_name);
+
+    // Check that the database has been found
+    if (!spectrum_json.is_open()) {
+        G4cerr << "IBDGen: [ERROR] spectrum database " << spectrum_name << " not found." << G4endl;
+        exit(-1);
+    } else {
+        G4cout << "IBDGen: [INFO] spectrum database " << spectrum_name << " found." << G4endl;
+    }
+
+    std::stringstream buffer;
+    buffer << spectrum_json.rdbuf();
+
+    json data = json::parse(buffer.str());
+
+    // Loop over the json
+    for (const auto &model : data["models"]) {
+        if (model["name"].get<string>() == model_name) {
+            G4cout << "IBDGen: [INFO] using model " << model_name << G4endl;
+
+            energy = model["energy"].get<std::vector<float>>();
+            flux = model["flux"].get<std::vector<float>>();
+
+            G4cout << "IBDGen: [INFO] spectrum read from database" << G4endl;
+
+            // Check if the flux vector and energy vector have the same size
```

Comment:
You should add another check, that the energy is monotonically increasing

Reply:
Added a check for this